wandb/sdk/service/streams.py

```diff
@@ -0,0 +1,305 @@
+        threading.Thread.__init__(self)
+        self.name = "StreamThr"
+        self._target = target
+        self._kwargs = kwargs
+        self.daemon = True
+        self.pid = 0
+
+    def run(self) -> None:
+        # TODO: catch exceptions and report errors to scheduler
+        self._target(**self._kwargs)
+
+
+class StreamRecord:
+    _record_q: "multiprocessing.Queue[pb.Record]"
+    _result_q: "multiprocessing.Queue[pb.Result]"
+    _relay_q: "multiprocessing.Queue[pb.Result]"
+    _iface: InterfaceRelay
+    _thread: StreamThread
+
+    def __init__(self) -> None:
+        self._record_q = multiprocessing.Queue()
+        self._result_q = multiprocessing.Queue()
+        self._relay_q = multiprocessing.Queue()
+        process = multiprocessing.current_process()
+        self._iface = InterfaceRelay(
+            record_q=self._record_q,
+            result_q=self._result_q,
+            relay_q=self._relay_q,
+            process=process,
+            process_check=False,
+            # use_router=False,
+        )
+
+    def start_thread(self, thread: StreamThread) -> None:
+        self._thread = thread
+        thread.start()
+        self._wait_thread_active()
+
+    def _wait_thread_active(self) -> None:
+        result = self._iface.communicate_status()
+        # TODO: using the default communicate timeout, is that enough? retries?
+        assert result
+
+    def join(self) -> None:
+        self._iface.join()
+        self._record_q.close()
+        self._result_q.close()
+        self._relay_q.close()
+        if self._thread:
+            self._thread.join()
+
+    def drop(self) -> None:
+        self._iface._drop = True
+
+    @property
+    def interface(self) -> InterfaceRelay:
+        return self._iface
+
+
+class StreamAction:
+    _action: str
+    _stream_id: str
+    _processed: Event
+    _data: Any
+
+    def __init__(self, action: str, stream_id: str, data: Any = None):
+        self._action = action
+        self._stream_id = stream_id
+        self._data = data
+        self._processed = Event()
+
+    def __repr__(self) -> str:
+        return f"StreamAction({self._action},{self._stream_id})"
+
+    def wait_handled(self) -> None:
+        self._processed.wait()
+
+    def set_handled(self) -> None:
+        self._processed.set()
+
+    @property
+    def stream_id(self) -> str:
+        return self._stream_id
+
+
+class StreamMux:
+    _streams_lock: threading.Lock
+    _streams: Dict[str, StreamRecord]
+    _port: Optional[int]
+    _pid: Optional[int]
+    _action_q: "queue.Queue[StreamAction]"
+    _stopped: Event
+
+    def __init__(self) -> None:
+        self._streams_lock = threading.Lock()
+        self._streams = dict()
+        self._port = None
+        self._pid = None
+        self._stopped = Event()
+        self._action_q = queue.Queue()
+
+    def _get_stopped_event(self) -> "Event":
+        # TODO: clean this up, there should be a better way to abstract this
```

Comment:
This refers to the Event logic we talked about? Do you want to track this one as well?

Reply:
Added to:
https://wandb.atlassian.net/browse/WB-7684